src/cdsetool/download.py

```diff
@@ -26,31 +30,53 @@
 from cdsetool.monitor import NoopMonitor, StatusMonitor
 from cdsetool.query import FeatureQuery
 
+MANIFEST_FILENAMES = {
+    "SENTINEL-1": "manifest.safe",
+    "SENTINEL-2": "manifest.safe",
+    "SENTINEL-3": "manifest.xml",
+}
 
-def download_feature(
-    feature, path: str, options: Union[Dict[str, Any], None] = None
-) -> Union[str, None]:
+
+def filter_files(manifest_file: str, pattern: str, exclude: bool = False) -> List[str]:
     """
-    Download a single feature
+    Filter a product's files, listed in its manifest, based on a given pattern.
 
-    Returns the feature ID
+    Returns a list of files paths within the product bundle.
+
+    If "exclude" is set to False, only files that match pattern are returned. If it's
+    set to True, only files that do not match pattern are returned.
     """
-    options = options or {}
-    log = _get_logger(options)
-    url = _get_feature_url(feature)
-    title = feature.get("properties").get("title")
-    temp_dir_usr = _get_temp_dir(options)
+    paths = []
+    xmldoc = etree.parse(manifest_file)
 
-    if not url or not title:
-        log.debug(f"Bad URL ('{url}') or title ('{title}')")
-        return None
+    if os.path.basename(manifest_file) == "manifest.safe":
+        data_obj_section_elem = xmldoc.find("dataObjectSection")
+        for elem in data_obj_section_elem.iterfind("dataObject"):  # type: ignore
+            path = Path(elem.find("byteStream/fileLocation").attrib["href"])  # type: ignore
+            paths.append(path)
 
-    filename = title + ".zip"
-    result_path = os.path.join(path, filename)
+    elif os.path.basename(manifest_file) == "manifest.xml":
+        namespaces = {"ns": "http://www.eumetsat.int/sip"}
+        data_section_elem = xmldoc.find("ns:dataSection", namespaces)
+        for elem in data_section_elem.iterfind("ns:dataObject", namespaces):  # type: ignore
+            path = Path(elem.find("ns:path", namespaces).text)  # type: ignore
+            # Remove prefix present for some files in S3 manifests
+            path = Path(*path.parts[-1:])
+            paths.append(path)
+
+    paths = [str(path) for path in paths if fnmatch.fnmatch(path, pattern) ^ exclude]
```

Comment:
Put the return on this line instead of redefining `paths`. The indentation level is 8 spaces on the line before so it's difficult to think this somehow belongs to the previous block, so I think the preceding blank line can be removed too.